### PR TITLE
NewClassTrees and MethodInvocation trees always have an element.

### PR DESCRIPTION
**Pull request by smillst**
Merge https://github.com/typetools/checker-framework/pull/4352 first.

### CODE REVIEW

File: checker/tests/nullness/java-unsound/Figure1.java [hunk 1]
```diff
@@ -1,5 +1,6 @@
 // Unsound only in Java 8, Java 9+ already gives an error
 // @below-java9-jdk-skip-test
+// @skip-test no need to test for the javac error.
```
Comment (mernst) [inline checker/tests/nullness/java-unsound/Figure1.java:3]: Should the file be removed (and likewise for the other newly skipped files)?

File: docs/manual/introduction.tex [hunk 2]
```diff
@@ -557,6 +557,8 @@
   which is a javac argument to halt compilation if a warning is issued.
 \item \<-AignoreInvalidAnnotationLocations>
   Ignore annotations in bytecode that have invalid annotation locations.
+\item \<-AnoWarnOnTypeCheckingHalt>
+  Don't issue the ``Javac errored; type checking halted.'' warning.
```
Comment (mernst) [inline docs/manual/introduction.tex:561]: Does type-checking still halt in this case?  (I assume so, but please make it explicit either way.)

When would a user want to pass this command-line argument?  If we don't have a compelling use-case, maybe this flag should not be added.

I see that two tests use it and they fail if it is removed, but I'm not sure why.
For instance, one fails this way:

```
org.checkerframework.framework.test.junit.InitializedFieldsValueTest > run[/home/mernst/research/types/checker-framework-fork-smillst-branch-no-null-element/framework/initialized-fields-value] FAILED
    java.lang.AssertionError: 3 out of 3 expected diagnostics were found.
    1 unexpected diagnostic was found:
    :0: warning: Javac errored; type checking halted.
    While type-checking /home/mernst/research/types/checker-framework-fork-smillst-branch-no-null-element/framework/tests/initialized-fields-value/ClassInitializer.java, /home/mernst/research/types/checker-framework-fork-smillst-branch-no-null-element/framework/tests/initialized-fields-value/ClassInitializer2.java, /home/mernst/research/types/checker-framework-fork-smillst-branch-no-null-element/framework/tests/initialized-fields-value/ClassInitializer2a.java, /home/mernst/research/types/checker-framework-fork-smillst-branch-no-null-element/framework/tests/initialized-fields-value/ClassInitializer3.java, /home/mernst/research/types/checker-framework-fork-smillst-branch-no-null-element/framework/tests/initialized-fields-value/ManualConstructor.java
```

but I am able to run `javac` on all of those files without problem.

File: framework/src/main/java/org/checkerframework/common/basetype/BaseTypeChecker.java [hunk 3]
```diff
@@ -757,7 +762,9 @@ public void typeProcessingOver() {
         for (BaseTypeChecker checker : getSubcheckers()) {
             checker.typeProcessingOver();
         }
-
+        if (javacErrored && parentChecker == null && !hasOption("noWarnOnTypeCheckingHalt")) {
+            messager.printMessage(Kind.WARNING, "Javac errored; type checking halted.");
```
Comment (mernst) [inline framework/src/main/java/org/checkerframework/common/basetype/BaseTypeChecker.java:766]: This message makes it sound like javac is buggy.  I think the point is that the program is not valid Java:  normal compilation without pluggable type-checking didn't succeed.  Please clarify the message.  Text like "javac issued an error" might be good, though from the user perspective javac also issues the pluggable type-checking errors, so the message should probably be even more explicit or detailed than that.  Another clarification is that *pluggable* type checking halted.

Also, please clarify what is the consequence to users, or why they care.  Might there be `.class` files generated that didn't get type-checked?  If javac erred, wouldn't users see other warning messages?

**Review by mernst — CHANGES_REQUESTED** [review]
I'm a bit confused about the notion of a javac error and what the consequences are for users.

File: checker/tests/nullness/java-unsound/Figure1.java (hunk 1 above)
Comment (smillst) [reply checker/tests/nullness/java-unsound/Figure1.java:3]: See reply: https://github.com/typetools/checker-framework/pull/4352/files#r587872848

**Comment by smillst** [comment]
Closed in favor of https://github.com/typetools/checker-framework/pull/4385.

**Pull request closed without merging**